File: backend-py/function/random_id.py
from snowflake_id import Generator

def is_pretty(numstr):
    if numstr.startswith('0'):
        return False
    for i in range(len(numstr) - 2):
        if numstr[i] == numstr[i+1] == numstr[i+2]:
            return False
        
    for i in range(len(numstr) - 3):
        s = numstr[i:i+4]
        if s in '0123456789' or s in '9876543210':
            return False
        try:
            nums = [int(c) for c in s]
            if nums == list(range(nums[0], nums[0]+4)):
                return False
            if nums == list(range(nums[0], nums[0]-4, -1)):
                return False
        except:
            continue
    return True

def generate_pretty_id():
    gen = Generator(datacenter_id=1, worker_id=1)
    while True:
        id_int = gen.generate()
        numstr = str(id_int)[-10:]
        if is_pretty(numstr):
            return numstr
        

File: backend/function/randomId/randomId.js
const Snowflake = require('snowflake-id').Snowflake;

function isPretty(numstr) {
  if (numstr.startsWith('0')) return false;
  for (let i = 0; i < numstr.length - 2; i++) {
    if (numstr[i] === numstr[i+1] && numstr[i] === numstr[i+2]) return false;
  }
  for (let i = 0; i < numstr.length - 3; i++) {
    const s = numstr.slice(i, i+4);
    if ('0123456789'.includes(s) || '9876543210'.includes(s)) return false;
    try {
      const nums = Array.from(s).map(Number);
      // Ascending
      if (nums.join('') === [...Array(4).keys()].map(k => nums[0]+k).join('')) return false;
      // Descending
      if (nums.join('') === [...Array(4).keys()].map(k => nums[0]-k).join('')) return false;
    } catch {}
  }
  return true;
}

function generatePrettyId() {
  const gen = new Snowflake({ mid: 1, offset: (2020-1970)*31536000*1000 });
  while (true) {
    const id = gen.generate().toString();
    const numstr = id.slice(-10);
    if (isPretty(numstr)) return numstr;
  }
}

module.exports = { generatePrettyId };
